pkg/predicate/predicate.go

```diff
@@ -202,3 +205,114 @@ func AddTypePredicate(extensionType string, predicates []predicate.Predicate) []
 	preds = append(preds, HasType(extensionType))
 	return append(preds, predicates...)
 }
+
+// MachineStatusHasChanged is a predicate deciding wether the status of a MCM's Machine has been changed.
+func MachineStatusHasChanged() predicate.Predicate {
+	statusHasChanged := func(oldObj runtime.Object, newObj runtime.Object) bool {
+		oldMachine, ok := oldObj.(*machinesv1alpha1.Machine)
+		if !ok {
+			return false
+		}
+		newMachine, ok := newObj.(*machinesv1alpha1.Machine)
+		if !ok {
+			return false
+		}
+		oldStatus := oldMachine.Status
+		newStatus := newMachine.Status
+
+		//Check the Node
+		if oldStatus.Node != newStatus.Node {
+			return true
+		}
+
+		//Check the CurrentStatus
+		if !equality.Semantic.DeepEqual(oldStatus.CurrentStatus, newStatus.CurrentStatus) {
+			return true
+		}
+
+		if !equality.Semantic.DeepEqual(oldStatus.LastOperation, newStatus.LastOperation) {
+			return true
+		}
+
+		//Check the Conditions
+		if !equality.Semantic.DeepEqual(oldStatus.Conditions, newStatus.Conditions) {
+			return true
+		}
+
+		return false
+	}
+
+	return statusChanged(statusHasChanged)
+}
+
+// MachineSetStatusHasChanged is a predicate deciding whether the status of a MCM's MachineSet has been changed.
+func MachineSetStatusHasChanged() predicate.Predicate {
+	statusHasChanged := func(oldObj runtime.Object, newObj runtime.Object) bool {
+		oldMachineSet, ok := oldObj.(*machinesv1alpha1.MachineSet)
+		if !ok {
+			return false
+		}
+		newMachineSet, ok := newObj.(*machinesv1alpha1.MachineSet)
+		if !ok {
+			return false
+		}
+		oldStatus := oldMachineSet.Status
+		newStatus := newMachineSet.Status
+
+		//Check the The number of actual replicas
+		if oldStatus.Replicas != newStatus.Replicas {
+			return true
+		}
+
+		//Check the number of pods that have labels matching the labels of the pod template of the replicaset.
+		if oldStatus.FullyLabeledReplicas != newStatus.FullyLabeledReplicas {
+			return true
+		}
+
+		//Check the number of ready replicas for this replica set.
+		if oldStatus.ReadyReplicas != newStatus.ReadyReplicas {
+			return true
+		}
+
+		//Check the number of available replicas for this replica set.
+		if oldStatus.AvailableReplicas != newStatus.AvailableReplicas {
+			return true
+		}
+
+		//Check LastOperation
+		if !equality.Semantic.DeepEqual(oldStatus.LastOperation, newStatus.LastOperation) {
+			return true
+		}
+
+		//Check the MachineSetConditions
+		if !equality.Semantic.DeepEqual(newStatus.Conditions, newStatus.Conditions) {
+			return true
+		}
+
+		//Check MachineSummary
+		if !equality.Semantic.DeepEqual(oldStatus.FailedMachines, newStatus.FailedMachines) {
+			return true
+		}
+
```

Comment:
Same question like above - which data here is dynamic that we have to keep it for the migration?

Reply:
Same comment as above